app/uk/gov/hmrc/gform/graph/Recalculation.scala

```diff
@@ -340,42 +399,41 @@ class Recalculation[F[_]: Monad, E](
             .applyTypeInfo(typeInfo)
         val hc = evaluationContext.headerCarrier
 
-        expressionResult.convertNumberToString.withStringResult(noStateChange(false)) { value =>
-          StateT[F, RecalculationState, Boolean] { s =>
-            val updS = s.update(evaluationResults + (expr, expressionResult))
-            def makeCall(): F[Boolean] = dataSource match {
-              case DataSource.Mongo(collectionName) => dbLookupCheckStatus(value, collectionName, hc)
-              case DataSource.Enrolment(serviceName, identifierName) =>
-                retrievals.enrolmentExists(serviceName, identifierName, value).pure[F]
-              case dd @ DataSource.DelegatedEnrolment(_, _) =>
-                retrievals.maybeGovermentGatewayId.fold(false.pure[F]) { governmentGatewayId =>
-                  delegatedEnrolmentCheckStatus(governmentGatewayId, dd, IdentifierValue(value), hc)
-                }
-              case DataSource.SeissEligible =>
-                seissEligibilityChecker(UtrEligibilityRequest(value), hc)
-            }
-
-            updS
-              .get(dataSource, value)
-              .fold(makeCall().map { res =>
-                (updS.add(dataSource, value, res), res)
-              })((updS, _).pure[F])
+        expressionResult.convertNumberToString.withStringResult(false.pure[F]) { value =>
+          exprMap.addOne((expr, expressionResult))
+          def makeCall(): F[Boolean] = dataSource match {
+            case DataSource.Mongo(collectionName) => dbLookupCheckStatus(value, collectionName, hc)
+            case DataSource.Enrolment(serviceName, identifierName) =>
+              retrievals.enrolmentExists(serviceName, identifierName, value).pure[F]
+            case dd @ DataSource.DelegatedEnrolment(_, _) =>
+              retrievals.maybeGovermentGatewayId.fold(false.pure[F]) { governmentGatewayId =>
+                delegatedEnrolmentCheckStatus(governmentGatewayId, dd, IdentifierValue(value), hc)
+              }
+            case DataSource.SeissEligible =>
+              seissEligibilityChecker(UtrEligibilityRequest(value), hc)
           }
+
+          (for {
+            a1 <- booleanExprCacheMap.get(dataSource)
+            a2 <- a1.get(value)
```

Comment:
Can we give `a1` and `a2` better more descriptive names?

Reply:
Done